### PR TITLE
[TOD][Teachers] Add TOD Teachers + base Parser agent

### PR DESCRIPTION
**Pull request by moyapchen**
Breaking up https://github.com/facebookresearch/ParlAI/pull/4176 into two diffs in the hope that makes these easier to review. 

This focuses on the *teachers used to train models*. Idea here is straightforward - Implement the Parser first, then use multiple inheritance to define the Teachers. 

This diff does 3 things:
1. Define the base parser Class
2. Define the Teachers which includes: 
    * metrics used in the teacher + its respective test
    * User + System Teacher
    * README 
3. Boilerplate for setting up tests

Helper agents + tests for teachers + agents in next diff in stack.

------------

NOTE: Diff commit history looks a little weird here since I accidentally merged something above into this stack, the reverted it. >.> (I did check this diff before merging and it's fine.)

### CODE REVIEW

File: parlai/core/tod/tod_agents.py [hunk 1]
```diff
@@ -0,0 +1,323 @@
...
+not need to be exposed outside of this file are prefixed with a single underscore ('_')
+"""
+
+from parlai.core.agents import Agent
+from parlai.core.message import Message
+from parlai.core.metrics import AverageMetric
+from parlai.core.params import ParlaiParser
+from parlai.core.opt import Opt
+from parlai.core.teachers import DialogTeacher
+from parlai.utils.distributed import is_distributed, get_rank, num_workers
```
Comment (moyapchen) [inline parlai/core/tod/tod_agents.py:21]: Ignore these lint failures - they are used in the next change in this stack.

**Review by skiingpacman — APPROVED** [review]
Left a few minor comments to take a look at, otherwise looks good.

File: parlai/core/tod/README.md [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+## Teachers for training (generative) models
+    * TodSystemTeacher
+    * TodUserSimulatorTeacher
+
+## Agents for Grounding
+For goal grounding for the User for simulation:
+    * TodGoalAgent
+        * Dumps goals as is from the dataset, possibly multiple per episode
+    * TodSingleGoalAgent
+        * Flattens goals such that a single one is used to seed a conversation. For datasets that include multiple goals per conversation, each individual goal is used as a seed.
```
[the remark this answers is not in the document]
Comment (moyapchen) [reply parlai/core/tod/README.md:38]: Yeah that's the idea; can clarify this.

File: parlai/core/tod/teacher_metrics.py [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+    def __init__(
+        self,
+        teacher_slots: Dict[str, str],
+        predicted_slots: Dict[str, str],
+        prefixes: Optional[List] = None,
+        shared: Dict[str, Any] = None,
+    ) -> None:
+        super().__init__(shared=shared)
+        self.prefixes = prefixes if prefixes else []
+        # jga and optionally Avg(jga,nlg_bleu)
```
[the remark this answers is not in the document]
Comment (moyapchen) [reply parlai/core/tod/teacher_metrics.py:32]: Oof, yeah, thanks for the catch.

File: parlai/core/tod/tod_agents.py [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+            "--episodes-randomization-seed",
+            type=int,
+            default=-1,
+            help="Randomize episodes in a predictable way (eg, for few shot). Set to -1 for no randomization. ",
+        )
+        parser.add_argument(
+            "--n-shot",
+            default=-1,
+            type=int,
+            help="Number of dialogues to keep for each of train/valid/test. -1 means all. Dialogues of lower numbers are strict subsets of larger numbers. Do not use in conjunction with `--percent-shot`. Use `--episodes-randomization-seed` to change seed. NOTE: Beware of using this flag when multitasking as this will apply to *all* datasets unless the ':' syntax for specifying per-dataset flags is used.",
```
[the remark this answers is not in the document]
Comment (moyapchen) [reply parlai/core/tod/tod_agents.py:58]: Yeah I was hoping for the linter to catch these, but they haven't been. =/

File: parlai/core/tod/README.md [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+---
+
+# Teachers + Agents Usage
+
+tl;dr Extend `TodStructuredDataParser` for your particular dataset and implement `generate_episode()` that converts the dataset into a list of episodes (`List[TodStructuredEpisode]`). Use multiple inheritence to generate teachers for training models. See files like `parlai/tasks/multiwoz_v22/agents.py` for an example. 
+
+## Overview of usage
+
+For a given dataset, extend `TodStructuredDataParser` and implement `generate_episodes()` and `get_id_task_prefix()`. The former of these is expected to do the data processing to convert a dataset to `List[TodStructuredEpisode]`. From here, multiple inheritance can be used to define Agents and Teachers that utilize the data.
```
[the remark this answers is not in the document]
Comment (moyapchen) [reply parlai/core/tod/README.md:17]: Oof good catch.

File: parlai/core/tod/tod_agents.py [hunk 6]
```diff
@@ -0,0 +1,323 @@
...
+            "--api-schemas",
+            type="bool",
+            default=False,
+            help="Preempt first turn with intents + required/optional parameters as key/value for given domain",
+        )
+        parser.add_argument(
+            "--api-jga-record",
+            type=bool,
+            default=True,
+            help="Should we save jga information per api schema?",
```
[the remark this answers is not in the document]
Comment (moyapchen) [reply parlai/core/tod/tod_agents.py]: Done.

**Pull request merged**